scipy/optimize/_differentialevolution.py

```diff
@@ -800,12 +886,74 @@ def _calculate_population_energies(self, population):
         return energies
 
     def _promote_lowest_energy(self):
-        # promotes the lowest energy to the first entry in the population
-        l = np.argmin(self.population_energies)
+        # swaps 'best solution' into first population entry
+
+        idx = np.arange(self.num_population_members)
+        feasible_solutions = idx[self.feasible]
+        if feasible_solutions.size:
+            # find the best feasible solution
+            idx_t = np.argmin(self.population_energies[feasible_solutions])
+            l = feasible_solutions[idx_t]
+        else:
+            # no solution was feasible, use 'best' infeasible solution, which
+            # will violate constraints the least
+            l = np.argmin(np.sum(self.constraint_violation, axis=1))
```

Comment:
This is never hit in tests -- the `if` is always True

Reply:
I can write a test for this, but I expect it will be very rare for this condition to be hit in practice, except if the constraints are really tough.

Reply:
addressed